metals/src/main/scala/scala/meta/internal/metals/testProvider/TestSuitesProvider.scala

```diff
@@ -1,165 +1,360 @@
+      symbol: Option[mtags.Symbol] = None
+  ): Seq[AddTestCases] =
+    for {
+      metadata <- index.getMetadata(path).toSeq
+      metadataEntry <- metadata.entries
+      testEntry <- index.getTestEntry(
+        metadataEntry.buildTarget,
+        metadataEntry.suiteName
+      )
+      if symbol.forall(_ == testEntry.symbol)
+      // if text document isn't defined try to fetch it from semanticdbs
+      doc <- textDocument.orElse(getSemanticDb(testEntry.symbol).map(_._2))
+    } yield {
+      val testClass = testEntry.testClass
+      val testCases = junitTestFinder.findTests(doc, path, testEntry.symbol)
+
+      if (testCases.nonEmpty) {
+        index.setHasResolvedChildren(path)
+      }
+
+      AddTestCases(
+        testClass.fullyQualifiedClassName,
+        testClass.className,
+        testCases.asJava
+      )
+    }
+
   /**
    * Find test suites for all build targets in current projects and update caches.
    */
   private def doRefreshTestSuites(): Future[Unit] =
-    if (clientConfig.isTestExplorerProvider()) Future {
+    if (isEnabled) Future {
       val buildTargetList = buildTargets.allBuildTargetIds.toList
+        // filter out JS and Native platforms
+        .filter(id =>
+          buildTargets
+            .scalaTarget(id)
+            .forall(_.scalaInfo.getPlatform == ScalaPlatform.JVM)
+        )
         .flatMap(buildTargets.info)
         .filterNot(_.isSbtBuild)
 
       val symbolsPerTarget = buildTargetList
-        .map(buildTarget =>
+        .map { buildTarget =>
           SymbolsPerTarget(
             buildTarget,
             buildTargetClasses.classesOf(buildTarget.getId).testClasses
           )
-        )
+        }
+
+      val deletedSuites = removeStaleTestSuites(symbolsPerTarget)
+      val addedEntries = getTestEntries(symbolsPerTarget)
 
-      // when test suite is deleted it has to be removed from cache
-      val classNamesFromBsp =
-        symbolsPerTarget.flatMap(_.testSymbols.values).toSet
-      val classNamesToRemove = cachedTestSuites.keySet -- classNamesFromBsp
-      classNamesToRemove.foreach { removedClass =>
-        cachedTestSuites.remove(removedClass)
+      // update cache
+      addedEntries.values.foreach {
+        _.foreach { case (entry, _) =>
+          index.put(entry)
+        }
       }
 
-      for {
-        SymbolsPerTarget(buildTarget, testSymbols) <- symbolsPerTarget
-        (symbol, fullyQualifiedClassName) <- testSymbols
-          .readOnlySnapshot()
-          .toList
-        // IMPORTANT this check is meant to check for class name, not a symbol
-        if !cachedTestSuites.contains(fullyQualifiedClassName)
-      } computeTestEntry(buildTarget, symbol, fullyQualifiedClassName)
+      val currentlyOpened = openedFiles().toSet
+      val addedTestCases = addedEntries.mapValues {
+        _.flatMap { case (entry, doc) =>
+          if (currentlyOpened.contains(entry.path))
+            findTestCases(entry.path, Some(doc))
```

Comment:
Since we never do anything about non opened files, there is no need to load semanticdb into memory for them.

Reply:
The semanticDB has been already loaded for the test suite in `computeTestEntry`